### ISSUE TITLE
Intermittent failures

### ISSUE THREAD

**Issue by dangershony** [hhblaze/DBreeze#52]
We rarely get this exception when writing to store `TableNotOperableException:Exception`.

We wonder what could be the reason and is that intermittent (applying a retry strategy will sort it?) 
Are there any other exceptions that can be considered retryable?

See discussion here
https://github.com/stratisproject/StratisBitcoinFullNode/pull/3243

The error
```
[2019-02-07 20:44:42.6119 36] ERROR: Stratis.Bitcoin.Features.Consensus.ProvenBlockHeaders.ProvenBlockHeaderStore+<SaveAsync>d__21.MoveNext Error saving the batch DBreeze.Exceptions.TableNotOperableException: DBreeze.TranJournal ---> DBreeze.Exceptions.DBreezeException: Rollback of the table "DBreeze.TranJournal" failed! ---> DBreeze.Exceptions.DBreezeException: Restore rollback file "/Deployment/NodeData/stratis/StratisMain/provenheaders/_DBreezeTranJrnl" failed! ---> System.IO.IOException: Remote I/O error
   at System.IO.FileStream.WriteNative(ReadOnlySpan`1 source)
   at System.IO.FileStream.FlushWriteBuffer()
   at System.IO.FileStream.Seek(Int64 offset, SeekOrigin origin)
   at System.IO.FileStream.set_Position(Int64 value)
   at DBreeze.Storage.FSR.Rollback()
   --- End of inner exception stack trace ---
   at DBreeze.LianaTrie.LTrieWriteCache.RollBack()
   at DBreeze.LianaTrie.LTrieRootNode.RollBack()
   --- End of inner exception stack trace ---
   at DBreeze.LianaTrie.LTrieRootNode.RollBack()
   at DBreeze.LianaTrie.LTrie.RollBack()
   --- End of inner exception stack trace ---
   at DBreeze.Transactions.TransactionsCoordinator.Commit(Int32 transactionThreadId)
   at DBreeze.Transactions.Transaction.Commit()
   at Stratis.Bitcoin.Features.Consensus.ProvenBlockHeaders.ProvenBlockHeaderRepository.<>c__DisplayClass16_0.<PutAsync>b__0() in H:\InternalTestnet_Data\Sources\StratisBitcoinFullNode\src\Stratis.Bitcoin.Features.Consensus\ProvenBlockHeaders\ProvenBlockHeaderRepository.cs:line 160
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
--- End of stack trace from previous location where exception was thrown ---
   at Stratis.Bitcoin.Features.Consensus.ProvenBlockHeaders.ProvenBlockHeaderStore.SaveAsync() in H:\InternalTestnet_Data\Sources\StratisBitcoinFullNode\src\Stratis.Bitcoin.Features.Consensus\ProvenBlockHeaders\ProvenBlockHeaderStore.cs:line 276
[2019-02-07 20:44:48.3362 36] INFO: Stratis.Bitcoin.Utilities.AsyncLoop+<>c__DisplayClass16_0+<<StartAsync>b__0>d.MoveNext FlushChain stopping.
[2019-02-07 20:44:50.1569 43] INFO: Stratis.Bitcoin.FullNode.<StartPeriodicLog>b__76_0 ======Node stats====== 02/07/2019 20:44:50
```

Thanks.

**Comment by hhblaze** [comment]
As I understand you have Linux/Windows Docker/Non-Docker Azure SSD/HDD environment and probably sometimes you got not enough Azure Disk IO's that strikes System.IO.FileStream.WriteNative?
If it is so, my first idea, that retry strategy on Disk Operations inside of DBreeze.FSR is not a good idea. 
May be somewhere on a higher level (program that uses DBreeze). 
Also may be it is not a bad idea to buy more Disk IO's from Azure.

**Comment by hhblaze** [comment]
Though, TableNotOperableException will not give you ability after such error, integrating it on a higher abstract level than Dbreeze is not a good idea. So may be you can clone FSR.cs integrate retries in all disk operations (Read, Seeks, Writes) and test a bit on your environment. Then we can go on discussion.

**Comment by dangershony** [comment]
This comment still applies? (we use netcore)
https://github.com/hhblaze/DBreeze/blob/master/DBreeze/Storage/FSR.cs#L25

**Comment by dangershony** [comment]
in this case `TableNotOperableException` was a recoverable error.

**Comment by hhblaze** [comment]
I have published version 1.93 where applied that FSR fix to .NetStandard (etc) versions, wait for NuGet. 
You are definitely entering DBreeze hall of fame.
Reopen issue if necessary.

**Comment by hhblaze** [comment]
> In this case `TableNotOperableException` was a recoverable error.

Of course, but settings that flag makes engine not usable until restart. 
So, please, test variations of retry scenarios or tell me how I can help you, then we will decide what to do further.
